HPCC-14535 Fetch all available ESPs and verify ESP is active

rpastrana opened the pull request:
- Fetches all available wsfs instances
- Verifies given wsfs is accessible

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
@wangkx please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f51178cca16f9a494b5e2050279bc37b6c8f508c
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.346 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

wangkx reviewed (commented):
Please see my comments.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b927709c5be82efee3e645b8633a885e4f86763f
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.443 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,136 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+{
+    StringBuffer key;
+    CriticalBlock b(espURLcrit);
+    espUrlsLastHeartBeat.setValue(key.set(url).str(), aliveTime);
```
I needed to persist url.

richardkchapman replied:
I don't understand this answer. The copy into key does not make any sense to me unless you are trying to translate nullptr to "" (but it's an inefficient way to do that...)

rpastrana commented:
@wangkx please review.

wangkx commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -222,16 +222,22 @@ static time_t contactWsFS(const char * espurl)
-    Owned<IClientEchoDateTimeResponse> result = server.EchoDateTime(req);
+    try
+    {
+		Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
```
@rpastrana Wrong indent.

rpastrana commented:
@wangkx please review again

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b906db387a52ec92daa1c586f3dbd263244c7a51
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.34 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented:
@rpastrana looks fine now.

rpastrana commented:
@richardkchapman please merge

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,141 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+        const char * key =  ((const char *)iter.query().getKey());
+        time_t  lastAlive = *(espUrlsLastHeartBeat.getValue(key));
+
+        if (now.getSimple() - lastAlive < 60 * 60)
```
You check once an hour? That seems a little infrequent...

rpastrana replied:
Polling intervals are always difficult, the right answer is usually to make it configurable, however based on your overall recommendation, this issue becomes moot.

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,141 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+
+    try
+    {
+        Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
```
How long does this take to time out if the server is NOT available? And is that a problem?

rpastrana replied:
In practice it's in the low seconds, but the fact we can't control the connect|read, etc timeouts on the generated ESP clients is an issue I'm looking into. I want to expose connection level configuration options.

richardkchapman commented:
@rpastrana A couple of questions, but also a more general point. I don't really understand why there is a need to be tracking availability / timestamps etc. The bug we are trying to fix is that it doesn't fail over when the server it previously connected to goes away... seems like all we need to do is (a) when first connecting, loop through all listed servers until we find one that responds and (b) repeat that process any time the one we talked to before doesn't respond.

What am I missing?

rpastrana commented:
@richardkchapman well, the concept was to avoid pinging the ESP service as much as possible while maintaining a "reasonably" spaced heartbeat test.

rpastrana commented:
@richardkchapman please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 72300175ba8d0e328a500f7d217109631e8521dc
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.279 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+    {
+        PROGLOG("Could not contact WsFS: '%s'",espurl);
+    }
+    return 0;
```
should say return false

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+        if (result->getResult())
+            return true;
+    }
+    catch(...)
```
Will leak any IException that was thrown...

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+        const char * currentUrl = availableWsFS.item(index);
+        if (contactWsFS(currentUrl, wu))
+            return currentUrl;
+        //else remove it from the list?
```
Should NOT remove from the list - if a user has no rights on a given ESP it does not mean that other users don't (and this list is static so shared by all users)

richardkchapman commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
-    return espurl.get();
-}
 
+    const char * wsfsurl = getNextAliveWsFSURL(wu);
```
Bit of a shame to start from beginning of list each time rather than last successful connection (though saving last successful connection per user might be a bit of a pain).

richardkchapman commented:
@rpastrana Some comments

rpastrana committed:
HPCC-14535 Fetch all configured WsFS and verify service is active

- Catches exceptions in WSFS method invocation
- Creates list of all configured WsFS services
- Attempts to contact WsFS before providing that target URL
- Cleans up error messages

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@richardkchapman changes based on latest review squashed in.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 02a9f50a695a0b00430a8fed1163a04dbcf5c696
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.39 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@richardkchapman new Jira based on client request timeout: https://track.hpccsystems.com/browse/HPCC-17178

richardkchapman commented:
@afishbeck @wangkx please re-review

afishbeck commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,129 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
```
@richardkchapman we have really got to find a better way than storing user credentials in an easily extracted manner within workunits.  Especially if we ever start using credentials that are valid for services outside of HPCC functionality.

rpastrana replied:
I think @RussWhitehead is looking into that matter https://track.hpccsystems.com/browse/HPCC-17172

afishbeck replied:
Ok, I commented on that issue.  Hopefully we can get rid of these entirely.

rpastrana replied:
Agreed. It seems the ECL Plugin logic shouldn't have to deal with credentials at all, either the environment executing the logic (eclide, etc) should provide the connection logic as well.

RussWhitehead replied:
I have proposed using Public/Private key files and OpenSSL to manage password encryption. See https://track.hpccsystems.com/browse/HPCC-17172

afishbeck reviewed (commented):
A few comments.

rpastrana commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,129 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+                {
                     Owned<IPropertyTreeIterator> iter3 = iter2->query().getElements("AuthenticateFeature");
-                    ForEach(*iter3) {
+                    ForEach(*iter3)
```
Legacy code, looks to be based on the assumption that looking for eclwatch alone wasn't guaranteed to always contain ws_fs??? 
Looking for service="ws_fs" will always yield false until the day somebody decides to split wsfs from eclwatch...

afishbeck replied:
EclWatch will always contain ws_fs unless we make a major (very unlikely) change in the future.  All services can be run on their own, although as you imply it may not be likely for this one.

The way authorization features ended up being done in the environment.xml is a bit weird, but the authorization isn't the best indicator of the service type, and even worse, another service may decide it needs to check ws_fs permissions even though it isn't the ws_fs service.  But again the setup config mechanism is weird.

rpastrana replied:
Agreed on all these points, and your suggestion is the obvious cleanest implementation. But no matter which way we go, we're making assumptions (WsFS will always be part of ecwatch, wsfs auth always associated with wsfs). It does't seem worthwhile to change the legacy implementation, but I'll go ahead a change it.

rpastrana replied:
The above suggestion forces a few more string compares, but saves an entire set of iterations over Auth entries, definitely worthwhile change.

rpastrana commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,129 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
                                 StringBuffer espname;
                                 if (iter1->query().getProp("@name",espname)) {
                                     StringBuffer espinst;
                                     if (iter1->query().getProp("Instance[1]/@computer",espinst)) {
```
Good point

richardkchapman commented:
@rpastrana I think this is back with you at present

rpastrana committed:
HPCC-14535 Fetch all available ESPs and verify ESP is active

- Fetches all available wsfs instances
- Verifies given wsfs is accessible

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -198,63 +198,132 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+
+        if (env.get())
+        {
+            StringBuffer wsFSUrl;
```
extracted from the inner loops to avoid re-initialization of the SB objects.

rpastrana commented:
@afishbeck please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 561edf93b39ad5c7df84aa104804b37072124d9d
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.267 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -273,53 +273,56 @@ static const char *getAccessibleEspServerURL(const char *param, IConstWorkUnit *
-                    Owned<IPropertyTreeIterator> iter3 = iter2->query().getElements("AuthenticateFeature");
-                    ForEach(*iter3)
+                    espBindingIter->query().getProp("@service",wsFSUrl.clear());
+                    if (wsFSUrl.length() && (stricmp(wsFSUrl.str(),"EclWatch")==0 || stricmp(wsFSUrl.str(),"ws_fs")==0))
```
Not a big deal, but I find strieq() easier to read.

Also this is why occasionally a more generic string name is good... if its being used for multiple purposes.  This name is no longer self describing.

afishbeck reviewed (approved):
@rpastrana looks better.

Still wish we could cache the list.

rpastrana commented:
@richardkchapman please merge.
